src/core/http/request_handler.rs

```diff
@@ -95,24 +100,38 @@ pub fn update_response_headers(
 }
 
 #[tracing::instrument(skip_all, fields(otel.name = "graphQL", otel.kind = ?SpanKind::Server))]
-pub async fn graphql_request<T: DeserializeOwned + GraphQLRequestLike>(
+pub async fn graphql_request<T: DeserializeOwned + GraphQLRequestLike + Hash + Send>(
     req: Request<Body>,
     app_ctx: &AppContext,
     req_counter: &mut RequestCounter,
 ) -> Result<Response<Body>> {
     req_counter.set_http_route("/graphql");
     let req_ctx = Arc::new(create_request_context(&req, app_ctx));
+    let headers = req.headers().clone();
     let bytes = hyper::body::to_bytes(req.into_body()).await?;
     let graphql_request = serde_json::from_slice::<T>(&bytes);
     match graphql_request {
         Ok(mut request) => {
-            let _ = request.parse_query();
-            let mut response = request.data(req_ctx.clone()).execute(&app_ctx.schema).await;
-
-            response = update_cache_control_header(response, app_ctx, req_ctx.clone());
-            let mut resp = response.into_response()?;
-            update_response_headers(&mut resp, &req_ctx, app_ctx);
-            Ok(resp)
+            if !app_ctx.blueprint.server.batch_execution && !is_query(&mut request) {
+                Ok(execute_query(&app_ctx, &req_ctx, request).await?)
+            } else {
+                let operation_id = OperationId::from(&request, &headers);
```

Comment:
- Implement `Hash` for `GraphQLRequestLike`
- Implement `is_query` for `GraphQLRequestLike`